File: Python/SA.py
"""
*模拟退火算法
*用于解决旅行商问题（TSP）
"""

from time import time
from copy import deepcopy
from math import sqrt, exp
from random import seed, randint, random

file_path: str = r'D:\Code\Project2\TSPExample_62128'  # 测试文件路径

T0 = 50000  # 起始温度
T_end = 1e-8  # 截止温度
q = 0.98  # 降温速率
L = 1000  # 迭代次数
N = 52  # 城市数量

city_list = []  # 城市路径

city_pos = []  # 城市坐标


def io():
    global N, city_pos
    with open(file_path, "r") as f:
        N = int(f.readline())
        for i in range(N):
            data = f.readline().split(" ")
            city_pos.append([float(data[1]), float(data[-1])])
        if f.readline() != "EOF\n":
            raise EOFError
    print("数据加载成功")


def distance(city1: tuple, city2: tuple) -> float:
    """
    计算城市间距离
    :param city1: 城市1的坐标
    :param city2: 城市2的坐标
    :return: 城市间距离
    """
    x1 = city1[0]
    x2 = city2[0]
    y1 = city1[1]
    y2 = city2[1]

    dis = sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2)
    return dis


def path_len(cities: list) -> float:
    """
    计算路径长度
    :param cities: 城市路径
    :return: 路径长度
    """
    path = 0
    for i in range(N - 1):
        index1 = cities[i]
        index2 = cities[i + 1]

        dis = distance(city_pos[index1 - 1], city_pos[index2 - 1])
        path += dis

    first_index = cities[0]
    last_index = cities[N - 1]
    last_dis = distance(city_pos[first_index - 1], city_pos[last_index - 1])
    path += last_dis
    return path


def init():
    """
    初始化城市路径
    """
    global city_list
    for i in range(N):
        city_list.append(i)


def create_new():
    """
    创造新的解
    """
    global city_list
    index1 = randint(0, N - 1)
    index2 = randint(0, N - 1)

    city_list[index1 - 1], city_list[index2 - 1] = city_list[index2 - 1], city_list[index1 - 1]


def main():
    global city_list
    seed(time())

    T = T0
    count = 0

    start = time()  # 开始计时
    io()
    init()
    while T > T_end:
        for i in range(L):
            city_list_copy = deepcopy(city_list)
            create_new()

            f1 = path_len(city_list)
            f2 = path_len(city_list_copy)

            dE = f1 - f2
            if dE > 0:  # 若遭遇上坡
                if exp(-dE / T) < random():  # 若拒绝接受新解
                    city_list = deepcopy(city_list_copy)
        print(f"当前温度温度T0={T:.2f}，最优路径长度为：{path_len(city_list):.2f}")
        T *= q
        count += 1
    end = time()  # 停止计时
    duration = end - start

    print(f"模拟退火算法，初始温度T0={T0:.2f}，降温系数q={q:.2f}")
    print(f"每个温度迭代{L:d}次，共降温{count:d}次")
    print(f"得到的TSP最优路径为:")

    for i in range(N - 1):
        print(f"{city_list[i]}--->", end='')
    print(f"{city_list[N - 1]}")
    print(f"最优路径长度为：{path_len(city_list):.2f}")

    hours = int(duration // 3600)
    minutes = int((duration - 3600 * hours) // 60)
    seconds = duration - 3600 * hours - 60 * minutes
    print(f"程序运行耗时：{hours:d}时 {minutes:d}分 {seconds:.2f}秒")


if __name__ == '__main__':
    main()

File: C++/SA.cpp
// 模拟退火算法
#include <iostream>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <math.h>
#define T0 50000 // 初始温度
#define T_end (1e-8)  // 最低温度
#define q 0.98 // 降温速率
#define L 1000 // 链长
#define N 31 // 城市数量
int city_list[N]; // 用于存放一个解

// 中国31个城市坐标
double city_pos[N][2] = {
{1304, 2312}, {3639, 1315}, {4177, 2244}, {3712, 1399},
{3488, 1535}, {3326, 1556}, {3238, 1229}, {4196, 1004},
{4312, 790}, {4386, 570}, {3007, 1970}, {2562, 1756},
{2788, 1491}, {2381, 1676}, {1332, 695},
{3715, 1678}, {3918, 2179}, {4061, 2370},
{3780, 2212}, {3676, 2578}, {4029, 2838},
{4263, 2931}, {3429, 1908}, {3507, 2367},
{3394, 2643}, {3439, 3201}, {2935, 3240},
{3140, 3550}, {2545, 2357}, {2778, 2826},
{2370, 2975}
};

//函数声明
double distance(double*, double*); // 计算两个城市距离
double path_len(int*); // 计算路径长度
void init(); // 初始化函数
void create_new(); // 产生新解

double distance(double* city1, double* city2) {
	double x1 = *city1;
	double x2 = *city2;
	double y1 = *(city1+1);
	double y2 = *(city2+1);
	double dis = sqrt((x1-x2)*(x1-x2)+(y1-y2)*(y1-y2));
	return dis;
}

double path_len(int* arr) {
	double path = 0;
	int index = *arr;
	for (int i = 0; i < N - 1; i++) {
		int index1 = *(arr + i);
		int index2 = *(arr + i + 1);
		double dis = distance(city_pos[index1-1], city_pos[index2-1]);

		path += dis;
	}
	int last_index = *(arr + N - 1);
	int first_index = *arr;
	double last_dis = distance(city_pos[last_index - 1], city_pos[first_index - 1]);

	path += last_dis;
	return path;
}

void init() {
	for (int i = 0; i < N; i++)
		city_list[i] = i + 1;
}

void create_new() {
	double r1 = ((double)rand() / (RAND_MAX + 1.0));
	double r2 = ((double)rand() / (RAND_MAX + 1.0));
	int pos1 = (int)(N * r1);
	int pos2 = (int)(N * r2);
	int temp = city_list[pos1];
	city_list[pos1] = city_list[pos2];
	city_list[pos2] = temp;
}

int main() {
	srand((unsigned)time(NULL));
	time_t start, finish;
	start = clock();
	double T = T0;
	int count = 0;
	init();
	int city_list_copy[N];
	double f1, f2, df;
	double r;

	while (T > T_end) {
		for (int i = 0; i < L; i++) {
			memcpy(city_list_copy, city_list, N * sizeof(int));
			create_new();
			f1 = path_len(city_list_copy);
			f2 = path_len(city_list);
			df = f2 - f1;
			if (df >= 0) {
				r = ((double)rand()) / RAND_MAX;
				if (exp(-df / T) <= r) {
					memcpy(city_list, city_list_copy, N * sizeof(int));
				}
			}
		}
		T *= q;
		count++;
	}
	finish = clock();
	double duration = ((double)(finish - start))/ CLOCKS_PER_SEC;

	printf("模拟退火算法，初始温度T0=%.2f，降温系数q=%.2f，\n每个温度迭代%d次，共降温%d次，\n得到的TSP最优路径为：\n",(double)T0, q, L, count);

	for (int i = 0; i < N - 1; i++) {
		printf("%d--->", city_list[i]);
	}
	printf("%d\n", city_list[N - 1]);
	double len = path_len(city_list);
	printf("最优路径长度为：%lf\n", len);
	printf("\n\n\n程序运行耗时：%lfs\n", duration);
	return 0;
}

